src/NetTopologySuite.Lab/Hull/ChiShape.cs

```diff
@@ -0,0 +1,151 @@
+﻿using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using NetTopologySuite.Geometries;
+using NetTopologySuite.Operation.Linemerge;
+using NetTopologySuite.Triangulate;
+using NetTopologySuite.Triangulate.QuadEdge;
+
+namespace NetTopologySuite.Hull
+{
+    public class ChiShape
+    {
+        private readonly Geometry _geom;
+        private readonly double _tolerance;
+
+        public ChiShape(Geometry geom, double tolerance)
+        {
+            _geom = geom;
+            _tolerance = tolerance;
+        }
+
+        /// <summary>
+        /// Generate the Delaunay triangulation of the set of input points P;
+        /// (2) Remove the longest exterior edge from the triangulation such that:
+        /// (a) the edge to be removed is longer than the tolerance parameter;
+        /// and (b) the exterior edges of the resulting triangulation form the boundary of a simple polygon;
+        /// (3) Repeat 2. as long as there are more edges to be removed (4) Return the polygon formed by the exterior edges of the triangulation
+        /// </summary>
+        /// <returns>Convex hull if tolerance is higher than the maximum edge length, simple polygon if tolerance is low.</returns>
+        /// <exception>If no concavehull can be generated throws not implemented exception.</exception>
+        public Geometry GetResult()
+        {
+            var subdiv = BuildDelaunay();
+            var tris = ExtractTriangles(subdiv);
+            var hull = ComputeHull(tris);
+            return hull;
+        }
+
+        private static IList<QuadEdgeTriangle> ExtractTriangles(QuadEdgeSubdivision subdiv)
+        {
+            var qeTris = QuadEdgeTriangle.CreateOn(subdiv);
+            return qeTris;
+        }
+
+
+        private Geometry ComputeHull(IList<QuadEdgeTriangle> tris)
+        {
+            var quadEdgeTriangleList = new SortedList<QuadEdge, QuadEdgeTriangle>(new QuadEdgeComparer());
```

Comment:
After going through everything else, I'm not sure `SortedList<TKey, TValue>` is the right thing to use here.  I'm favoring an `AlternativePriorityQueue<TPriority, TData>` with a `Dictionary<TKey, TValue>` on the side.

- The way we would have to contort our usage of `SortedList<TKey, TValue>` to eliminate an `O(n²)` bit is unnatural.  We only extract values on one end of the sorted list, which is what priority queues optimize for.
- The criteria we should be using for `.ContainsKey` is different from the criteria we use for sorting, which is incompatible with `SortedList<TKey, TValue>`.

Reply:
We will leave this as an exercise for the viewers.